pycloudlib/oci/utils.py

```diff
@@ -84,14 +82,126 @@ def get_subnet_id_by_name(
     return subnets[0].id
 
 
+def _get_subnet_features(
+    subnet: oci.core.models.Subnet,
+) -> Dict[str, Any]:
+    """
+    Get the availability domain, private, and networking type of the subnet.
+
+    Args:
+        subnet: The subnet model to get the features of.
+
+    Returns:
+        A dictionary containing the following keys:
+        - availability_domain: The availability domain of the subnet.
+        - private: Whether the subnet is private.
+        - networking_type: The networking type of the subnet
+
+    """
+    availability_domain = subnet.availability_domain
+    private = subnet.prohibit_internet_ingress
+    has_ipv4_cidr_block = subnet.cidr_block not in (None, "<null>")
+    has_ipv6_cidr_block = subnet.ipv6_cidr_block not in (None, "<null>")
+    networking_type = None
+    if has_ipv4_cidr_block and not has_ipv6_cidr_block:
+        networking_type = NetworkingType.IPV4
+    elif not has_ipv4_cidr_block and has_ipv6_cidr_block:
+        networking_type = NetworkingType.IPV6
+    elif has_ipv4_cidr_block and has_ipv6_cidr_block:
+        networking_type = NetworkingType.DUAL_STACK
+    else:
+        log.warning(
+            "Unable to determine networking type for subnet %s [%s]",
+            subnet.display_name,
+            subnet.id,
+        )
+    return {
+        "availability_domain": availability_domain,
+        "private": private,
+        "networking_type": networking_type,
+    }
+
+
+def _subnet_is_compatible(
+    subnet: oci.core.models.Subnet,
+    availability_domain: str,
+    networking_config: NetworkingConfig,
+) -> bool:
+    """
+    Check if the subnet is compatible with the given restrictions.
+
+    For each restriction, the following must be true:
+    availability_domain:
+    - if the subnet has an availability domain, it must match the given availability domain
+    - otherwise, it is compatible
```

Comment:
It seems like there are more possibilities than are described here.

* the subnet has an AD
  * the subnet's AD matches the given AD --> compatible
  * the subnet's AD doesn't match the given AD -> incompatible
* the subnet does not have an AD -> compatible

It's a bit ambiguous which condition the word "otherwise" applies to -- having an AD, or matching the AD.